### PR TITLE
luminous: cls_bucket_list fails causes cascading osd crashes

### PR DESCRIPTION
**Pull request by smithfarm**
http://tracker.ceph.com/issues/24630

### CODE REVIEW

**Comment by smithfarm** [comment]
@yehudasa Have a look at the conflict resolution?

File: src/cls/rgw/cls_rgw.cc [hunk 1]
```diff
@@ -317,12 +317,20 @@ static void split_key(const string& key, list<string>& vals)
   }
 }
 
+static string escape_str(const string& s)
+{
+  int len = escape_json_attr_len(s.c_str(), s.size());
+  std::string escaped(len, 0);
+  escape_json_attr(s.c_str(), s.size(), escaped.data());
```
Comment (cbodley) [inline src/cls/rgw/cls_rgw.cc]: the non-const overload of `std::string::data()` is a c++17 feature, so this doesn't build in luminous. but there's a duplicate of this function [later in the file](https://github.com/ceph/ceph/blob/luminous/src/cls/rgw/cls_rgw.cc#L994) that uses a variable-length array instead, so we should just use that one

we should remove that duplicate definition in master too!

Comment (smithfarm) [reply src/cls/rgw/cls_rgw.cc]: @cbodley The duplicate `escape_str` definition is removed by this commit e71fba0c6f5b3b0572b5136840cf0ed3c9186569 in master, but thanks to your comment I now see what to do. Thanks!

**Comment by smithfarm** [comment]
https://github.com/smithfarm/ceph/tree/wip-24630-luminous is on 65cbf91153 but this PR is still on 7064ea1ef4b2d4c1e1d2a5557af47935e0024830 - waiting for github to catch up!

**Pull request closed without merging**